transports/quic/src/connection.rs

```diff
@@ -257,63 +200,168 @@ impl Connection {
-                dir: quinn_proto::Dir::Bi,
-            }) => ConnectionEvent::StreamAvailable,
-            quinn_proto::Event::Stream(quinn_proto::StreamEvent::Opened {
-                dir: quinn_proto::Dir::Bi,
-            }) => ConnectionEvent::StreamOpened,
-            quinn_proto::Event::ConnectionLost { reason } => {
-                ConnectionEvent::ConnectionLost(ConnectionError::from(reason).into())
+        };
+        inner.substreams.insert(substream_id, Default::default());
+        let substream = Substream::new(substream_id, self.state.clone());
+
+        Poll::Ready(Ok(substream))
+    }
+
+    fn poll_outbound(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        let mut inner = self.state.lock();
+        let substream_id = match inner.connection.streams().open(quinn_proto::Dir::Bi) {
+            Some(id) => {
+                inner.poll_outbound_waker = None;
+                id
             }
-            quinn_proto::Event::Stream(quinn_proto::StreamEvent::Finished { id }) => {
-                ConnectionEvent::StreamFinished(id)
+            None => {
+                inner.poll_outbound_waker = Some(cx.waker().clone());
+                return Poll::Pending;
             }
-            quinn_proto::Event::HandshakeDataReady => ConnectionEvent::HandshakeDataReady,
-            quinn_proto::Event::Stream(quinn_proto::StreamEvent::Opened {
-                dir: quinn_proto::Dir::Uni,
-            })
-            | quinn_proto::Event::Stream(quinn_proto::StreamEvent::Available {
-                dir: quinn_proto::Dir::Uni,
-            })
-            | quinn_proto::Event::DatagramReceived => {
-                unreachable!("We don't use datagrams or unidirectional streams.")
+        };
+        inner.substreams.insert(substream_id, Default::default());
+        let substream = Substream::new(substream_id, self.state.clone());
+        Poll::Ready(Ok(substream))
+    }
+
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let mut inner = self.state.lock();
+        if inner.connection.is_drained() {
+            return Poll::Ready(Ok(()));
+        }
+
+        for substream in inner.substreams.keys().cloned().collect::<Vec<_>>() {
+            let _ = inner.connection.send_stream(substream).finish();
+        }
+
+        if inner.connection.streams().send_streams() == 0 && !inner.connection.is_closed() {
+            inner
+                .connection
+                .close(Instant::now(), From::from(0u32), Default::default())
+        }
+        drop(inner);
+
+        loop {
+            match ready!(self.poll_event(cx)) {
+                Some(quinn_proto::Event::ConnectionLost { .. }) => return Poll::Ready(Ok(())),
+                None => return Poll::Ready(Err(Error::EndpointDriverCrashed)),
+                _ => {}
```

Comment:
Should we put an `unreachable!` here if we never expect this case? I'm afraid that if we for some reason hit this case in the future we will eternally busy loop.

Also do we not have to worry about `Stream(_)` events?

Reply:
I misunderstood. It's not that we don't expect this case, it's that we don't care about this case and it's safe to drop.

We don't worry about `Stream(_)` events so we drop them.

Reply:
This could be made more explicit with `mem::drop` but I am not sure it is worth it :)